### PR TITLE
Add maybe_set and try_with

### PR DESCRIPTION
**Pull request by Zoxc**
These also turned out to be useful in rustc. [Although it's not very pretty.](https://github.com/Zoxc/rust/blob/rayon-queries/src/librustc/ty/context.rs#L1205)

### CODE REVIEW

**Comment by alexcrichton** [comment]
Since `try_with` can be implemented in terms of `is_set` I'm tempted to leave it out for now, but can something like `unset` be added instead of `maybe_set`?

**Comment by Zoxc** [comment]
I don't need something like `unset` in `rustc`. I just though these functions were useful as a pair.

**Comment by alexcrichton** [comment]
Is `maybe_with` never called with `None`?

**Comment by Zoxc** [comment]
Only inside a closure passed to `try_with` which gets called with `None`.

**Comment by alexcrichton** [comment]
Er doesn't that mean that rustc need `unset`?

**Comment by Zoxc** [comment]
No. It is already unset in that case.

**Comment by alexcrichton** [comment]
Er well ok let me rephrase. I would prefer to avoid adding too many convenience methods and instead stick to a small set of primitives. I don't think either of these methods fall into that category though?
